### PR TITLE
refactor settings, add AwardYear scope

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -124,4 +124,15 @@ def need_authentication?
     (Rails.env.staging? || Rails.env.production?) &&
     controller_name != "healthchecks"
   end
+
+  def load_award_year_and_settings
```
Comment (matthewford) [inline app/controllers/application_controller.rb]: Assignment Branch Condition size for load_award_year_and_settings is too high. [15.65/15]

File: app/models/form_answer.rb [hunk 2]
```diff
@@ -38,6 +37,7 @@ class FormAnswer < ActiveRecord::Base
   begin :associations
     belongs_to :user
     belongs_to :account
+    belongs_to :award_year
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 3]
```diff
@@ -90,15 +90,15 @@ def secondary
   end
 
   begin :scopes
-    scope :for_award_type, -> (award_type) { where award_type: award_type }
-    scope :for_year, -> (year) { where award_year: year }
+    scope :for_award_type, -> (award_type) { where(award_type: award_type) }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 4]
```diff
@@ -90,15 +90,15 @@ def secondary
...
+    scope :for_year, -> (year) { joins(:award_year).where(award_years: { year: year }) }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 5]
```diff
@@ -90,15 +90,15 @@ def secondary
...
     scope :shortlisted_with_no_certificate, -> { where("1 = 0") }
     scope :winners, -> { where("1 = 0") }
     scope :unsuccessful, -> { where(state: %w(not_recommended reserved)) }
   end
 
   begin :callbacks
-    before_save :set_award_year, unless: :award_year?
+    before_save :set_award_year, unless: :award_year
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/services/notifiers/email_notification_service.rb [hunk 6]
```diff
@@ -24,32 +25,32 @@ def run
     end
   end
 
-  def shortlisted_notifier
+  def shortlisted_notifier(award_year)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Unused method argument - `award_year`. If it's necessary, use `_` or `_award_year` as an argument name to indicate that it won't be used. You can also write as `shortlisted_notifier(*)` if you want the method to accept any arguments but don't care about them.

File: app/services/notifiers/email_notification_service.rb [hunk 7]
```diff
@@ -24,32 +25,32 @@ def run
...
     User.shortlisted.each do |user|
       Users::NotifyShortlistedMailer.notify(user).deliver_later!
     end
   end
 
-  def not_shortlisted_notifier
+  def not_shortlisted_notifier(award_year)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Unused method argument - `award_year`. If it's necessary, use `_` or `_award_year` as an argument name to indicate that it won't be used. You can also write as `not_shortlisted_notifier(*)` if you want the method to accept any arguments but don't care about them.

File: spec/services/notifiers/email_notification_service_spec.rb [hunk 8]
```diff
@@ -134,5 +134,22 @@
...
+                            kind: kind,
+                            settings: previous_award_year.settings)
+
+      form_answer.update_attribute(:award_year, previous_award_year)
+
+      mailer = double(deliver_later!: true)
+      expect(Users::UnsuccessfulFeedbackMailer).to receive(:notify).with(form_answer.id) { mailer }
+
+      described_class.run
+3
```
Comment (matthewford) [inline spec/services/notifiers/email_notification_service_spec.rb]: Inconsistent indentation detected.

File: spec/support/shared_contexts/admin_feedback_pdf_file_checks.rb [hunk 9]
```diff
@@ -44,7 +44,7 @@
   end
 
   let(:award_general_information) do
-    "#{FeedbackPdfs::General::DrawElements::AWARD_GENERAL_INFO_PREFIX} #{form_answer.award_year}"
+    "#{FeedbackPdfs::General::DrawElements::AWARD_GENERAL_INFO_PREFIX} #{form_answer.award_year.year}"
```
Comment (matthewford) [inline spec/support/shared_contexts/admin_feedback_pdf_file_checks.rb]: Line is too long. [102/100]

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 10]
```diff
@@ -3,22 +3,22 @@
 describe "FeedbackPdfs::Base" do
   let!(:form_answer_2014_innovation) do
     FactoryGirl.create :form_answer, :submitted, :innovation,
-                        award_year: 2014
+                        award_year: AwardYear.where(year: 2014).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 11]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2015_innovation) do
     FactoryGirl.create :form_answer, :submitted, :innovation,
-                        award_year: 2015
+                        award_year: AwardYear.where(year: 2015).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 12]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2014_trade) do
     FactoryGirl.create :form_answer, :submitted, :trade,
-                        award_year: 2014
+                        award_year: AwardYear.where(year: 2014).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 13]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2015_trade) do
     FactoryGirl.create :form_answer, :submitted, :trade,
-                        award_year: 2015
+                        award_year: AwardYear.where(year: 2015).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: db/migrate/20150408160359_add_award_year_id_to_form_answers.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+class AddAwardYearIdToFormAnswers < ActiveRecord::Migration
+  def up
+    add_column :form_answers, :award_year_id, :integer, null: false
```
Comment (jaleszek) [inline db/migrate/20150408160359_add_award_year_id_to_form_answers.rb]: it would break migration if any record exists.

Comment (dreamfall) [inline db/migrate/20150408160359_add_award_year_id_to_form_answers.rb]: Yes, we'll need to recreate DB.

File: app/models/award_year.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+  has_many :form_answers
+  has_one :settings, inverse_of: :award_year, autosave: true
+
+  after_create :create_settings
+
+  AVAILABLE_YEARS = [2016, 2017, 2018, 2019]
+
+  # +1 here is to match URN number of assosiated forms
+  def self.current
```
Comment (jaleszek) [inline app/models/award_year.rb]: I though we don't want to have application awarding year associated with calendar year ? How it works when 2 awarding campaigns happens during one year or one exceeds the New Year and covers 2 calendar years ?

File: db/migrate/20150408160359_add_award_year_id_to_form_answers.rb (hunk 14 above)
Comment (jaleszek) [inline db/migrate/20150408160359_add_award_year_id_to_form_answers.rb]: ok if it's not an issue, that's fine

File: db/migrate/20150408153214_create_award_years.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+class CreateAwardYears < ActiveRecord::Migration
+  def change
+    create_table :award_years do |t|
+      t.integer :year
```
Comment (jaleszek) [inline db/migrate/20150408153214_create_award_years.rb]: `null: false` ?

Comment (dreamfall) [inline db/migrate/20150408153214_create_award_years.rb]: yeah

**Comment by jaleszek** [comment]
@dreamfall worth to rebase

**Comment by jaleszek** [comment]
@dreamfall something wrong with these tests

**Comment by jaleszek** [comment]
specs fail in master

**Comment by RuslanHamidullin** [comment]
I will check mailer collab spec later - as it on master

File: app/models/award_year.rb [hunk 17]
```diff
@@ -0,0 +1,29 @@
+class AwardYear < ActiveRecord::Base
+  validates :year, presence: true, uniqueness: true
+
+  has_many :form_answers
+  has_one :settings, inverse_of: :award_year, autosave: true
+
+  after_create :create_settings
+
+  AVAILABLE_YEARS = [2016, 2017, 2018, 2019]
```
Comment (matthewford) [inline app/models/award_year.rb]: Should this be a fixed list? It'll have to be updated each year - should be auto calculated?

Comment (dreamfall) [inline app/models/award_year.rb]: @matthewford I guess it can be changed once we'll do the proper switch between award years in admin area.

**Pull request merged**